### PR TITLE
Edit Graph

### PR DESCRIPTION
**Pull request by pldespaigne**
State issue number in your PR to close it when PR merged: issue #XXX
e.g. : "Resolve the bundle size ❤️: issue #1234"

### CODE REVIEW

File: libs/waterfall/src/lib/components/graph/label/label.component.ts [hunk 1]
```diff
@@ -10,6 +15,13 @@ import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
 })
 export class WaterfallGraphLabelComponent {
 
-  @Input() amount: number;
+  amount$ = this.shell.state$.pipe(
+    map(state => Object.values(state.waterfall.state.transfers).find(t => t.from === this.arrow.parentId && t.to === this.arrow.childId)?.amount ?? 0),
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/label/label.component.ts:19]:
```suggestion
    map(state => state.waterfall.state.transfers[`${this.arrow.parentId}->${this.arrow.childId}]?.amount ?? 0),
```

File: libs/waterfall/src/lib/components/graph/right/right.component.ts [hunk 2]
```diff
@@ -16,18 +18,49 @@ export class WaterfallGraphRightComponent implements OnChanges {
...
 
+  @Output() addChild = new EventEmitter<string>();
+  @Output() addSibling = new EventEmitter<string>();
   @Output() handleSelect = new EventEmitter<string>();
 
   @HostBinding('class.nodrag') nodrag = true;
   @HostBinding('class.selected') selectedClass = false;
 
+  amount$ = this.shell.state$.pipe(
+    map(state => state.waterfall.state.transfers[this.right.id]?.amount ?? 0),
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/right/right.component.ts:32]: transfers keys are `from->to`  is `this.right.id` like this ?

File: libs/waterfall/src/lib/components/graph/vertical/vertical.component.ts [hunk 3]
```diff
@@ -14,6 +17,30 @@ export class WaterfallGraphVerticalComponent {
...
   @Input() vertical: VerticalNode;
   @Input() selected: boolean;
+  @Input() editMode = false;
 
+  @Output() addChild = new EventEmitter<string>();
+  @Output() addSibling = new EventEmitter<string>();
   @Output() handleSelect = new EventEmitter<string>();
+
+  amount$ = this.shell.state$.pipe(
+    map(state => state.waterfall.state.transfers[this.vertical.id]?.amount ?? 0),
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/vertical/vertical.component.ts:27]: transfers keys are from->to is this.vertical.id like this ?

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 4]
```diff
@@ -407,15 +407,451 @@ function getArrow(parent: Node, child: Node, transfers: TransferState[]) {
...
+  graph.forEach(node => {
+    node.children.forEach(childId => {
+      parentIndex[childId] ||= [];
+      parentIndex[childId].push(node.id);
+    });
+  });
+
+  graph.forEach(node => {
+    if (node.type === 'source') {
+      const source: WaterfallSource = {
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:717]: use createWaterfallSource

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 5]
```diff
@@ -407,15 +407,451 @@ function getArrow(parent: Node, child: Node, transfers: TransferState[]) {
...
+        };
+        rights.push(right);
+      }
+    }
+  });
+
+  return { rights, sources };
+}
+
+
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:837]: I suggest to create and use functions like `createXXX` to make code more readable and reduce number of line of this file

File: apps/waterfall/waterfall/src/app/dashboard/title/waterfall/waterfall.component.html [hunk 6]
```diff
@@ -1,16 +1,9 @@
...
-      ></waterfall-graph>
-    </ng-container>
-  </ng-container>
-</ng-container>
+
+<waterfall-graph
+  *ngIf="(rights$ | async) as rights; else loading"
+  [rights$]="rights$"
+  [waterfall$]="waterfall$"
+></waterfall-graph>
```
[the remark this answers is not in the document]
Comment (pldespaigne) [reply apps/waterfall/waterfall/src/app/dashboard/title/waterfall/waterfall.component.html:6]: TBD later

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**